lib/study_cleanup_tools.rb

```diff
@@ -0,0 +1,159 @@
+              puts "#{ws_project}/#{ws_name} successfully deleted"
+            else
+              puts "skipping #{ws_project}/#{ws_name} because it was created by #{ws_owner}"
+            end
+          rescue => e
+            puts "Unable to delete #{ws_project}/#{ws_name} due to: #{e.class.name} - #{e.message}"
+          end
+        end
+      end
+      true
+    end
+  end
+
+  ##
+  # security safeguards to prevent accidental deletes of production/staging workspaces and studies
+  ##
+
+  # ensure commands cannot be run on deployed hosts
+  #
+  # * *raises*
+  #   - (ArgumentError) => if hostname matches DISALLOWED_HOSTS
+  #
+  # * *returns*
+  #   - (TrueClass) => true if hostname passes validation
+  def self.validate_hostname!
+    current_host = Socket.gethostname
+    if current_host =~ DISALLOWED_HOSTS
+      raise ArgumentError.new("#{current_host} is not a permitted host for running StudyCleanupTools methods")
+    else
+      true
+    end
+  end
+
+  # ensure commands cannot be run against protected billing projects for production/staging
+  #
+  # * *raises*
+  #   - (ArgumentError) => if billing project matches DISALLOWED_BILLING_PROJECTS, or is not the configured project for this instance
+  #
+  # * *returns*
+  #   - (TrueClass) => true if billing project passes validation
+  def self.validate_billing_project!(project_name)
+    if !permit_billing_project?(project_name)
+      raise ArgumentError.new("#{project_name} is not a permitted billing project for running StudyCleanupTools methods")
+    else
+      true
+    end
+  end
+
+  # ensure commands cannot be run in production/staging environments
+  # can be run manually in development if allow_dev_env is set to true
+  #
+  # * *raises*
+  #   - (ArgumentError) => if environment is not test/pentest, or is development and allow_dev_env = false
+  #
+  # * *returns*
+  #   - (TrueClass) => true if environment passes validation
+  def self.validate_environment!(allow_dev_env = false)
+    environments = %w(test pentest)
+    allow_dev_env ? environments << 'development' : nil
+    if !environments.include?(Rails.env)
+      raise ArgumentError.new("#{Rails.env} is not a permitted environment for running StudyCleanupTools methods")
+    else
+      true
+    end
+  end
+
+  # ensure commands cannot be run in production/staging environments
+  # can be run manually in development if allow_dev_env is set to true
+  #
+  # * *raises*
+  #   - (ArgumentError) => if this is not a continuous integration run (governed by ENV['CI'] == true)
+  #
+  # * *returns*
+  #   - (TrueClass) => true if CI run passes validation
+  def self.validate_continuous_integration!
+    if !ENV['CI'] || ENV['CI'].blank?
+      raise ArgumentError.new("This is not a continuous integration test run, ENV['CI'] is not true")
+    else
+      true
+    end
+  end
+
+  private
+
+  # is requested project permitted to run a command
+  def self.permit_billing_project?(project_name)
```

Comment:
this seems like it should just be inlined in validate_billing_project.  `validate_billing_project` has no other logic, so I don't think you're gaining testability

Reply:
It's actually used in two places - one in the validation as the check, and also in `destroy_all_studies_and_workspaces` as a final check on whether or not to delete a study.  But given the other refactoring, I think this will all get cleaned up.